### PR TITLE
Dns response 2198 v5

### PR DESCRIPTION
**Pull request by StianHB**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues/) ticket:[2198](https://redmine.openinfosecfoundation.org/issues/2198)

Describe changes:
- Created the dns_response keyword that allows matching of byte strings on dns responses
- Updated the user-docs to reflect the new keyword

Incorporated changes from #3004, #3006 and #3057

### CODE REVIEW

File: doc/userguide/rules/dns-keywords.rst [hunk 1]
```diff
@@ -22,6 +22,17 @@ used in a rule all contents following it are affected by it.  Example:
 The dns_query keyword affects all following contents, until pkt_data
 is used or it reaches the end of the rule.
 
+dns_response
+------------
+
+The **dns_response** keyword allows you to match IPs in A record responses.
```
Comment (norg) [inline doc/userguide/rules/dns-keywords.rst]: only for A records? If I see it correct AAAA records are also supported?

Comment (StianHB) [reply doc/userguide/rules/dns-keywords.rst]: Yeah, I changed the code from the last PR. Its for all response types now and byte matching so no IP string either.

Comment (jasonish) [reply doc/userguide/rules/dns-keywords.rst]: Please fix.

File: src/app-layer-dns-common.c [hunk 2]
```diff
@@ -851,11 +851,6 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
         case DNS_RECORD_TYPE_A:
         {
             if (datalen == 0 || datalen == 4) {
-                //PrintRawDataFp(stdout, data, ntohs(head->len));
-                //char a[16];
-                //PrintInet(AF_INET, (const void *)data, a, sizeof(a));
-                //SCLogInfo("A %s TTL %u", a, ntohl(head->ttl));
-
```
Comment (jasonish) [inline src/app-layer-dns-common.c]: Please re-add this comment out code. It doesn't appear to be on consequence to your PR.

File: src/app-layer-dns-common.c [hunk 3]
```diff
@@ -870,10 +865,6 @@ const uint8_t *DNSReponseParse(DNSState *dns_state, const DNSHeader * const dns_
         case DNS_RECORD_TYPE_AAAA:
         {
             if (datalen == 0 || datalen == 16) {
-                //char a[46];
-                //PrintInet(AF_INET6, (const void *)data, a, sizeof(a));
-                //SCLogInfo("AAAA %s TTL %u", a, ntohl(head->ttl));
-
```
Comment (jasonish) [inline src/app-layer-dns-common.c]: Same here.

File: src/detect-engine-dns.c [hunk 4]
```diff
@@ -178,19 +281,29 @@ int PrefilterTxDnsQueryRegister(SigGroupHead *sgh, MpmCtx *mpm_ctx)
...
+}
+
+
 int DetectEngineInspectDnsRequest(ThreadVars *tv,
-        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
-        const Signature *s, const SigMatchData *smd,
-        Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+                                  DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+                                  const Signature *s, const SigMatchData *smd,
+                                  Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
```
Comment (jasonish) [inline src/detect-engine-dns.c]: Please remove these changes in white space.

File: src/detect-engine-dns.c [hunk 5]
```diff
@@ -178,19 +281,29 @@ int PrefilterTxDnsQueryRegister(SigGroupHead *sgh, MpmCtx *mpm_ctx)
...
 }
 
-int DetectEngineInspectDnsResponse(ThreadVars *tv,
-        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
-        const Signature *s, const SigMatchData *smd,
-        Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
+int DetectEngineInspectDnsResponseGeneric(ThreadVars *tv,
+                                    DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+                                    const Signature *s, const SigMatchData *smd,
+                                    Flow *f, uint8_t flags, void *alstate, void *txv, uint64_t tx_id)
```
Comment (jasonish) [inline src/detect-engine-dns.c]: White space.

**Comment by jasonish** [comment]
Replaced by: https://github.com/OISF/suricata/pull/3111

**Pull request closed without merging**